engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/source/KafkaGenericNodeSourceFactory.scala

```diff
@@ -0,0 +1,72 @@
+import pl.touk.nussknacker.engine.kafka.serialization.KafkaDeserializationSchemaFactory
+
+import scala.reflect.ClassTag
+
+
+class KafkaGenericNodeSourceFactory[T: ClassTag](deserializationSchemaFactory: KafkaDeserializationSchemaFactory[T],
+                                                 timestampAssigner: Option[TimestampWatermarkHandler[T]],
+                                                 formatter: RecordFormatter,
+                                                 processObjectDependencies: ProcessObjectDependencies,
+                                                 variableProvider: Option[VariableProvider[T]])
+  extends BaseKafkaSourceFactory(deserializationSchemaFactory, timestampAssigner, formatter, processObjectDependencies)
+    with SingleInputGenericNodeTransformation[FlinkSource[T]] {
+
+  override type State = Nothing
+
+  override def initialParameters: List[Parameter] = Parameter[String](KafkaGenericNodeSourceFactory.TopicParamName)
+    .copy(editor = Some(DualParameterEditor(
+      simpleEditor = StringParameterEditor,
+      defaultMode = DualEditorMode.RAW))
+    )  :: Nil
+
+  override def contextTransformation(context: ValidationContext, dependencies: List[NodeDependencyValue])(implicit nodeId: ProcessCompilationError.NodeId)
+  : NodeTransformationDefinition = {
+    case TransformationStep(Nil, _) => NextParameters(initialParameters)
+    case TransformationStep((KafkaGenericNodeSourceFactory.TopicParamName, _)::Nil, None) => FinalResults(finalCtx(context, dependencies))
+  }
+
+  private def finalCtx(context: ValidationContext, dependencies: List[NodeDependencyValue])(implicit nodeId: NodeId): ValidationContext = {
+    val name = dependencies.collectFirst {
+      case OutputVariableNameValue(name) => name
+    }.get
+
+    variableProvider
+      .map(_.validationContext(context, name))
+      .getOrElse(context.withVariable(OutputVar.customNode(name), Typed[T]).getOrElse(context))
+  }
+
+  override def implementation(params: Map[String, Any], dependencies: List[NodeDependencyValue], finalState: Option[Nothing]): FlinkSource[T] = {
+    val topic = params(KafkaGenericNodeSourceFactory.TopicParamName).asInstanceOf[String]
+    val kafkaConfig = KafkaConfig.parseProcessObjectDependencies(processObjectDependencies)
+    createSource(List(topic), kafkaConfig)
+  }
+
+  override protected def createSource(topics: List[String], kafkaConfig: KafkaConfig): KafkaSource[T] = {
+    val preparedTopics = topics.map(KafkaUtils.prepareKafkaTopic(_, processObjectDependencies))
+    val deserializationSchema = deserializationSchemaFactory.create(topics, kafkaConfig)
+    new KafkaSource(preparedTopics, kafkaConfig, deserializationSchema, timestampAssigner, formatter) {
+      override def initContext(processId: String, taskName: String): InitContextFunction[T] =
+        variableProvider.map(_.initContext(processId, taskName)).getOrElse(super.initContext(processId, taskName))
```

Comment:
I don't like this mixin of composition and inheritance. Let's have a fallback to some DefaultVariableProvider or sth like this.

Reply:
Fixed. See `FlinkIntermediateRawSource`